debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -0,0 +1,159 @@
+        this.pollInterval = pollInterval;
+        this.maxBatchSize = maxBatchSize;
+        this.queue = new LinkedBlockingDeque<>(maxQueueSize);
+        this.metronome = Metronome.sleeper(pollInterval, Clock.SYSTEM);
+        this.loggingContextSupplier = loggingContextSupplier;
+    }
+
+    public static class Builder<T> {
+
+        private Duration pollInterval;
+        private int maxQueueSize;
+        private int maxBatchSize;
+        private Supplier<LoggingContext.PreviousContext> loggingContextSupplier;
+
+        public Builder<T> pollInterval(Duration pollInterval) {
+            this.pollInterval = pollInterval;
+            return this;
+        }
+
+        public Builder<T> maxQueueSize(int maxQueueSize) {
+            this.maxQueueSize = maxQueueSize;
+            return this;
+        }
+
+        public Builder<T> maxBatchSize(int maxBatchSize) {
+            this.maxBatchSize = maxBatchSize;
+            return this;
+        }
+
+        public Builder<T> loggingContextSupplier(Supplier<LoggingContext.PreviousContext> loggingContextSupplier) {
+            this.loggingContextSupplier = loggingContextSupplier;
+            return this;
+        }
+
+        public ChangeEventQueue<T> build() {
+            return new ChangeEventQueue<T>(pollInterval, maxQueueSize, maxBatchSize, loggingContextSupplier);
+        }
+    }
+
+    public void enqueue(T record) {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            queue.put(record);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Placed source record '{}' into queue", record);
+            }
+        } catch (InterruptedException e) {
+            LOGGER.debug("received interrupt request");
+            // clear the interrupted status
+            Thread.interrupted();
+        } finally {
+            previousContext.restore();
+        }
+    }
+
+    public List<T> poll() {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            LOGGER.debug("polling records...");
+            List<T> records = new ArrayList<>();
+            final Timer timeout = Threads.timer(Clock.SYSTEM, Temporals.max(pollInterval, ConfigurationDefaults.RETURN_CONTROL_INTERVAL));
+            while (!timeout.expired() && queue.drainTo(records, maxBatchSize) == 0) {
+                throwProducerFailureIfPresent();
+
+                try {
+                    LOGGER.debug("no records available yet, sleeping a bit...");
+                    // no records yet, so wait a bit
+                    metronome.pause();
+                    LOGGER.debug("checking for more records...");
+                } catch (InterruptedException e) {
+                    // we've been requested to stop polling
+                    Thread.interrupted();
```

Comment:
DO we really want to clear interrupted() flag? General question - when thread interrupt can happen? Is it triggered by Connect?

Reply:
Good question, I've copied that from the PG connector, but I think it's actually not useful to do it. And indeed, the MongoDB and MySQL implementations don't do it. Instead, I'll change the signature of `ChangeEventQueue#poll()` to throw `InterruptedException`. That way, the exception can bubble up all the way to KC which can deal with it.

> when thread interrupt can happen? Is it triggered by Connect?

That's my assumption, yes. Although I didn't see the place where it's called after a quick look into KC. But I could imagine it gets called if a call to `SourceTask#poll()` keeps blocking for a while after `stop()` has been invoked.

Reply:
Maybe it was a bug in the code and what was meant was `Thread.interrupt()` to set the `interrupted` flag back again

Reply:
So I don't think the polling loop will ever be interrupted, it seems KC (currently) relies on implementors to properly abort if requested to stop. I think we're fine, as we handle real thread interruption via `Metronome#sleep()` and also honor the request to stop by evaluating the `stop` flag.

I'll change `enqueue()` to properly honor thread interruption.